### PR TITLE
feat(builders): implement custom builders

### LINKED ISSUE #3: how to: custom builder
official builder tutorial: https://angular.io/guide/cli-builder

reported issue with all the findings:

https://github.com/nrwl/nx/issues/2211

### PR DESCRIPTION
**Pull request by Hotell**
- demonstrate how to implement custom builders
- adds thorough docs how to implement custom builders (will be published as blog post)
- use custom builders in architect targets
- add custom schematic to generate custom builders
- add custom npm script similar to `workspace-schematic` -> to generate nx custom builders `workspace-builders:generate`

Closes #3

### CODE REVIEW

File: tools/builders/command/index.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    child.stdout.on('data', (data) => {
+      context.logger.info(data.toString());
+    });
+    child.stderr.on('data', (data) => {
+      context.logger.error(data.toString());
+    });
+
+    context.reportStatus(`Done.`);
+    child.on('close', (code) => {
+      context.logger.info('DONE!!!');
```
Comment (jukben) [inline tools/builders/command/index.ts:34]: !!! xD

File: tools/builders/file/file.impl.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  BuilderOutput,
+  createBuilder
+} from '@angular-devkit/architect';
+import { Observable, bindNodeCallback, of } from 'rxjs';
+import { catchError, map, tap } from 'rxjs/operators';
+import { json, getSystemPath, normalize } from '@angular-devkit/core';
+import { writeFile } from 'fs';
+import dateFormat from 'dateformat';
+
+// interface Schema extends TimestampBuilderSchema, json.JsonObject {}
```
Comment (jukben) [inline tools/builders/file/file.impl.ts]: nit: strip it?

File: tools/schematics/workspace-builder/index.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+// import { chain, externalSchematic, Rule } from '@angular-devkit/schematics';
```
Comment (jukben) [inline tools/schematics/workspace-builder/index.ts]: nit: strip it?

File: tools/schematics/workspace-builder/files/__name__/index.js__tmpl__ [hunk 4]
```diff
@@ -0,0 +1,7 @@
+const babelRegister = require('@babel/register');
+const config = require('../babel-register.config');
+
+babelRegister(config);
+
+// REQUIRED BUILDER API: re-export default as default
```
Comment (jukben) [inline tools/schematics/workspace-builder/files/__name__/index.js__tmpl__:6]: 👍

File: tools/schematics/workspace-builder/index.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    branchAndMerge(
+      chain([createFiles(options), updateBuildersSchema(options)])
+    ),
+    formatFiles(options)
+  ]);
+}
+
+function createFiles(options: NormalizedSchema): Rule {
+  const templateSource = apply(url('./files'), [
+    template({
```
Comment (jukben) [inline tools/schematics/workspace-builder/index.ts:41]: little bit magical for my taste 🧙‍♀️ I would maybe add a comment or two. What do you think?

**Review by jukben — APPROVED** [review]

File: tools/schematics/workspace-builder/index.ts (hunk 5 above)
Comment (Hotell) [reply tools/schematics/workspace-builder/index.ts:41]: well, it's just interpolating values to template content and template file names

image

File: tools/builders/file/file.impl.ts (hunk 2 above)
Comment (Hotell) [reply tools/builders/file/file.impl.ts]: will do!

File: tools/schematics/workspace-builder/index.ts (hunk 3 above)
Comment (Hotell) [reply tools/schematics/workspace-builder/index.ts]: will do!

File: tools/schematics/workspace-builder/index.ts (hunk 5 above)
Comment (Hotell) [reply tools/schematics/workspace-builder/index.ts:41]: I'll create docs for it (at least to reference library docs)

File: tools/schematics/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Schematics
```
Comment (Hotell) [inline tools/schematics/README.md:1]: is this sufficient @jukben  ? thx

Comment (jukben) [reply tools/schematics/README.md:1]: absolutely! Thanks a lot!

**Review by jukben — APPROVED** [review]

**Pull request merged**